### PR TITLE
Re-export 17 duplicated utilities from ultralytics

### PR DESCRIPTION
**Pull request by glenn-jocher**
Removes ~290 net lines of code duplicated in the `ultralytics` package. Each symbol was compared body-to-body against upstream and, where numerics were involved, checked with randomized trials before deletion.

### Deleted and re-exported

| file | symbols | source |
|---|---|---|
| `utils/general.py` | `Profile`, `get_latest_run`, `print_args`, `one_cycle`, `xywh2xyxy`, `xywhn2xyxy`, `xyxy2xywhn`, `segments2boxes`, `clip_boxes` | `ultralytics.utils.{ops,files,checks,torch_utils}` |
| `utils/torch_utils.py` | `model_info`, `is_parallel`, `initialize_weights`, `smart_amp_autocast` | `ultralytics.utils.torch_utils` |
| `utils/dataloaders.py` | `get_hash`, `img2label_paths`, `seed_worker`, `autosplit` | `ultralytics.data.{utils,build,split}` |
| `utils/dataloaders.py` | `exif_transpose` | `PIL.ImageOps` — ultralytics ships no such symbol |
| `utils/metrics.py`, `utils/segment/general.py` | `bbox_ioa`, `mask_iou`, `masks2segments` | `ultralytics.utils.{metrics,ops}` |

The imports keep the `# noqa: F401` re-export shape already used in this file, so the ~40 modules importing these names from `utils.*` need no edits. Orphans `import math` (`general.py`), `from pathlib import Path` (`torch_utils.py`) and `import hashlib` (`dataloaders.py`).

### Two load-bearing call-site edits

- **`models/yolo.py:205`** — upstream `model_info`'s second positional is `detailed`, not `verbose`. Left positional, `img_size` would have landed in `verbose`. Now `model_info(self, detailed=verbose, imgsz=img_size)`.
- **`segment/predict.py:180`** — upstream `masks2segments` defaults to `strategy="all"`; this repo assumed `"largest"`. The call now pins it. Without the pin `--save-txt` polygons silently switch from largest-contour to merged-contour.

`bbox_ioa` needs a 2-D `box1` upstream, so its one call site in `utils/augmentations.py` passes `np.array(box)[None]`. The result becomes `(1, n)` instead of `(n,)`; the only consumer is `(ioa < 0.30).all()`.

### Two functions deliberately kept local

Both now carry a `# Keep local (do not dedup)` comment so this doesn't get re-proposed:

- **`xyxy2xywh`** — upstream opens with `assert x.shape[-1] == 4`, but `models/common.py:970` calls it as `[xyxy2xywh(x) for x in pred]` where each `pred[i]` is `(n, 6)` = `xyxy, conf, cls`. The local version copies and writes only columns 0–3, preserving the trailing ones. Re-exporting would raise `AssertionError` on `results.xywh` / `results.xywhn` for every `torch.hub` user.
- **`clean_str`** — the regex character classes differ by exactly two characters: local strips `´` and keeps `` ` ``, upstream is the reverse. It feeds stream-source display paths and output filenames. yolov3 already carries this same keep-local marker.

### Cosmetic, user-visible

- Model summaries count leaf modules and gain thousands separators: `YOLOv5n summary: 214 layers, 1872157 parameters` → `125 layers, 1,872,157 parameters`. Parameters, gradients and **GFLOPs are unchanged** (verified n/s/m).
- `print_args` sorts arguments alphabetically and scrubs credentials from logged URLs.
- `Profile` times with `perf_counter` instead of `time.time` — affects printed ms only.
- `autosplit` gains `avif`/`heic`/`heif`/`jp2` and loses `pfm`, following ultralytics' `IMG_FORMATS`. Its one caller is the `data/xView.yaml` prep script.

### Validation

| check | result |
|---|---|
| `val.py --weights yolov5n.pt --data coco128.yaml --img 640` | `all 128 929 0.607 0.5 0.549 0.35` — identical to master |
| `segment/val.py --weights yolov5n-seg.pt --data coco128-seg.yaml --img 640` | all 8 metrics identical to master |
| `segment/predict.py --save-txt` | polygon `.txt` files **byte-identical** to master (`diff -r` clean) |
| `get_hash` on real paths | digest identical → existing `*.cache` files stay valid, confirmed by a train run reusing a master-written cache |
| `bbox_ioa` | 2000 randomized trials, max abs delta **0.0** |
| `train.py --epochs 1 --device cpu`, `detect.py` | pass |
| `pytest -m "not network"` | 20 passed |
| `ruff check` / `ruff format --check` | clean |

## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

### 🌟 Summary

🔄 Refactors YOLOv5 to reuse shared Ultralytics utilities, reducing duplicated code while preserving YOLOv5-specific behavior where needed.

### 📊 Key Changes

- Migrated common data, augmentation, metrics, segmentation, model, and training utilities to the `ultralytics` package.
- Updated xView dataset splitting to use `ultralytics.data.split.autosplit`.
- Replaced local EXIF handling with `PIL.ImageOps.exif_transpose`.
- Updated model information reporting to match the shared utility API.
- Moved segmentation helpers such as `masks2segments` and `mask_iou` to shared Ultralytics utilities.
- Removed duplicated implementations for hashing, label-path generation, random seeding, box conversions, profiling, AMP autocasting, model initialization, and model summaries.
- Kept select YOLOv5-specific functions locally where behavior differs, including `print_args`, `clean_str`, `xyxy2xywh`, and `print_args`.
- Adjusted bounding-box input handling in copy-paste augmentation and selected the largest contour when converting masks to segments.

### 🎯 Purpose & Impact

- 🧩 Improves consistency between YOLOv5 and the broader Ultralytics codebase.
- 🛠️ Reduces maintenance overhead by centralizing shared functionality.
- 🚀 Enables YOLOv5 to benefit from fixes and improvements made to common Ultralytics utilities.
- ✅ Preserves compatibility for YOLOv5-specific behavior that cannot be directly deduplicated.
- 📦 May affect downstream users importing removed helpers from YOLOv5’s local utility modules; those functions should now be imported from the corresponding `ultralytics` modules instead.

### CODE REVIEW

**Comment by UltralyticsAssistant** [comment]
👋 Hello @glenn-jocher, thank you for submitting a `ultralytics/yolov5` 🚀 PR! This automated message helps with the initial review; an engineer will assist with your contribution. Please review the following checklist:
- ✅ **Define a Purpose**: Clearly explain the purpose of your fix or feature in your PR description, and link to any [relevant issues](https://github.com/ultralytics/yolov5/issues). Ensure your commit messages are clear, concise, and adhere to the project's conventions.
- ✅ **Synchronize with Source**: Confirm your PR is synchronized with the `ultralytics/yolov5` `main` branch. If it's behind, update it by clicking the 'Update branch' button or by running `git pull` and `git merge main` locally.
- ✅ **Ensure CI Checks Pass**: Verify all Ultralytics [Continuous Integration (CI)](https://docs.ultralytics.com/help/CI) checks are passing. If any checks fail, please address the issues.
- ✅ **Update Documentation**: Update the relevant [documentation](https://docs.ultralytics.com/) for any new or modified features.
- ✅ **Add Tests**: If applicable, include or update tests to cover your changes, and confirm that all tests are passing.
- ✅ **Sign the CLA**: Please ensure you have signed our [Contributor License Agreement](https://docs.ultralytics.com/help/CLA) if this is your first Ultralytics PR by writing "I have read the CLA Document and I sign the CLA" in a new message.
- ✅ **Minimize Changes**: Limit your changes to the **minimum** necessary for your bug fix or feature addition. _"It is not daily increase but daily decrease, hack away the unessential. The closer to the source, the less wastage there is."_ — Bruce Lee
For more guidance, please refer to our [Contributing Guide](https://docs.ultralytics.com/help/contributing). Don't hesitate to leave a comment if you have any questions. Thank you for contributing to Ultralytics! 🚀

File: utils/dataloaders.py [hunk 1]
```diff
@@ -22,7 +21,11 @@
 import torchvision
 import yaml
 from PIL import ExifTags, Image, ImageOps
+from PIL.ImageOps import exif_transpose  # noqa: F401
 from torch.utils.data import DataLoader, Dataset, dataloader, distributed
+from ultralytics.data.build import seed_worker
+from ultralytics.data.split import autosplit  # noqa: F401
```
Comment (UltralyticsAssistant) [inline utils/dataloaders.py]: 💡 **MEDIUM**: This changes the public `autosplit()` default from YOLOv5's `DATASETS_DIR / "coco128/images"` to Ultralytics' `DATASETS_DIR / "coco8/images"` (and that upstream constant is configured independently of `YOLOV5_DATASETS_DIR`). Consequently, existing callers that invoke `utils.dataloaders.autosplit()` without `path` will split a different dataset or fail when only `coco128` is present. Keep a local compatibility wrapper with the YOLOv5 default, delegating to the upstream implementation.

File: utils/metrics.py [hunk 2]
```diff
@@ -7,7 +7,7 @@
 import matplotlib.pyplot as plt
 import numpy as np
 import torch
-from ultralytics.utils.metrics import box_iou, smooth
+from ultralytics.utils.metrics import box_iou, mask_iou, smooth
```
Comment (UltralyticsAssistant) [inline utils/metrics.py:10]: 💡 **MEDIUM**: `bbox_ioa` is removed from this module but is not re-exported here. This breaks the established public import path `from utils.metrics import bbox_ioa`; only `utils.augmentations` was updated to use the upstream location. Re-export it from this module to preserve downstream callers.

**Suggested change:**
```suggestion
from ultralytics.utils.metrics import bbox_ioa, box_iou, mask_iou, smooth
```

File: utils/segment/general.py [hunk 3]
```diff
@@ -111,37 +111,3 @@ def scale_image(im1_shape, masks, im0_shape, ratio_pad=None):
     if len(masks.shape) == 2:
         masks = masks[:, :, None]
     return masks
-
-
-def mask_iou(mask1, mask2, eps=1e-7):
```
Comment (UltralyticsAssistant) [inline utils/segment/general.py:116]: 💡 **MEDIUM**: Removing `mask_iou` and `masks2segments` without replacement removes both public exports from `utils.segment.general`. Updating the in-repository callers does not preserve downstream code using `from utils.segment.general import mask_iou, masks2segments`, despite the PR describing these utilities as re-exported. Import and re-export the upstream symbols from this module.

**Review by UltralyticsAssistant — COMMENTED** [review]
## 🔍 PR Review

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

The upstream substitutions are generally wired correctly, including the adjusted `model_info`, `bbox_ioa`, and `masks2segments` call signatures. However, this currently introduces three public-API regressions: two removed utility exports and a changed `autosplit()` default/configuration source.

💬 Posted 3 inline comments

**Comment by glenn-jocher** [comment]
Review pass applied. Three corrections, one of which changed the code's scope:

**`print_args` is now kept local.** Review caught that upstream computes `Path(file).resolve().relative_to(ROOT)` against the *ultralytics package* root, never this repo's — so it always takes the `ValueError` branch and falls back to `Path(file).stem`. The effect is that `train.py`, `segment/train.py` and `classify/train.py` all log the identical `train:` prefix, and the three predict scripts collapse the same way. That is a real loss of log disambiguation for 12 lines, so `print_args` joins `xyxy2xywh` and `clean_str` as a marked keep-local. Its body is restored byte-for-byte from master.

**The `exif_transpose` re-export is gone.** It was a second name for one callable, existing only so `models/common.py` wouldn't need an edit. Pointing that one call site at `PIL.ImageOps.exif_transpose` deletes both the shim and the duplicate name — smaller, and the right side of Delete > Add.

**Import ordering** in `utils/augmentations.py` fixed so the block stays isort-clean. (Ultralytics Actions independently pushed the identical fix, confirming it would have been reordered on merge.)

Also correcting a claim in the description above: the sentence *"The imports keep the `# noqa: F401` re-export shape ... so the ~40 modules importing these names from `utils.*` need no edits"* is true for `utils/general.py`, `utils/torch_utils.py` and `utils/dataloaders.py`, but **not** for `bbox_ioa`, `mask_iou` and `masks2segments` — those three are removed outright from `utils.metrics` / `utils.segment.general` and their in-repo call sites were edited instead. Nothing external imports them, so no shim is warranted, but the blanket claim was wrong.

Re-validated after the changes: `val.py` `all 128 929 0.607 0.5 0.549 0.35`, pytest 20 passed, `ruff check` + `ruff check --select I` clean, AutoShape round-trip returns the correct `results.files` and a `(4, 6)` `xywh` (exercising the kept-local `xyxy2xywh`), and `segment/predict:` still prints its full prefix.

**Comment by glenn-jocher** [comment]
Bot review addressed — one change applied, the rest declined with reasons.

### Applied: `autosplit` is no longer re-exported

The point about the default is correct and I have taken it further than suggested. Re-exporting kept `utils.dataloaders.autosplit` importable but silently repointed its zero-argument default from this repo's `DATASETS_DIR / "coco128/images"` to ultralytics' `SETTINGS`-configured datasets dir + `coco8/images`. A local wrapper restoring the old default would be a compatibility shim, which `AGENTS.md` rules out. Instead the re-export is dropped entirely, so the name is an explicit `ImportError` rather than a silently different one, and the single caller (`data/xView.yaml:91`) imports from `ultralytics.data.split` directly.

### Declined: re-exporting `bbox_ioa`, `mask_iou`, `masks2segments`

Same reasoning as #13834 and the same precedent: #13828 removed `bbox_iou` and `smooth_BCE` with no re-export. Adding `# noqa: F401` lines for symbols with no in-repo consumer is indirection that nothing calls — "three similar lines beat a helper nobody else calls" applies equally to a re-export nobody imports. All in-repo callers are updated.

The PR description previously claimed these three kept the re-export shape; that sentence was wrong and has been corrected in a comment above.

### Declined: the `github-code-quality[bot]` unused-import comments

All eight are false positives. Every flagged line is a deliberate `# noqa: F401` re-export that keeps the `utils.*` module API stable for the ~40 modules importing through it; acting on them would delete the mechanism this PR is built on. Two of them flag `intersect_dicts` and `file_date`/`file_size`, which were already re-exported on `master` before this PR touched anything. The `exif_transpose` and `print_args` comments were valid but are already resolved — both were removed in earlier commits on this branch for unrelated review reasons.

`ruff check .` (the linter this repo actually enforces, per `pyproject.toml`) passes clean, as does `ruff check --select I .`.

**Comment by UltralyticsAssistant** [comment]
🎉 Thank you, @glenn-jocher, for this excellent contribution—and congratulations on the merge!

> “Great things are done by a series of small things brought together.” — Vincent van Gogh

This PR brings shared Ultralytics utilities together across YOLOv5, reducing duplication, improving consistency, and making future maintenance and fixes more efficient while preserving YOLOv5-specific behavior. We appreciate the careful attention to compatibility and the continued work strengthening the broader Ultralytics ecosystem!
